src/pip/_internal/resolution/resolvelib/provider.py

```diff
@@ -75,6 +77,133 @@ def _get_with_identifier(
     return default
 
 
+def _extract_names_from_causes_and_parents(
+    causes: Iterable["PreferenceInformation"],
+) -> Set[str]:
+    """
+    Utility function to extract names from the causes and their parent packages
+
+    :params causes: An iterable of PreferenceInformation
+
+    Returns a set of strings, each representing the name of a requirement or
+    its parent package that was in causes
+    """
+    causes_names = set()
+    for cause in causes:
+        causes_names.add(cause.requirement.name)
+        if cause.parent:
+            causes_names.add(cause.parent.name)
+
+    return causes_names
+
+
+def _causes_with_conflicting_parent(
+    causes: Iterable["PreferenceInformation"],
+) -> List["PreferenceInformation"]:
+    """
+    Identifies causes that conflict because their parent package requirements
+    are not satisfied by another cause, or vice versa.
```

Comment:
I'm still struggling to understand the longic here. I'd suggest 2 improvements.

1. Clarify the above sentence - I don't really understand what "their parent package requirements are not satisfied by another cause" means, or why the parent package requirements *should* be satisfied by another cause.
2. Describe the data structure of "causes" here, explaining the elements that we're using. Something like "each cause contains a requirement specifier, and a "parent", which is the candidate which has that requirement. The list of causes only contains elements that have been detected by the resolver to be in conflict somehow." (I'm assuming that explanation is at least in part wrong, as it's not enough for me to be able to make sense of the function code...)